datafusion/functions-aggregate/src/first_last.rs

```diff
@@ -569,6 +573,13 @@ impl LastValueAccumulator {
             })
             .collect::<Vec<_>>();
 
+        // Order by indices for cases where the values are the same, we expect the last index
+        let indices: UInt64Array = (0..num_rows).map(|x| x as u64).collect();
+        sort_columns.push(SortColumn {
```

Comment:
This part is a bit concerning in terms of performance -- won't sorting by multiple columns (due to row indices being added, now there are no cases with a single column) be noticeably slower due to falling back to lexicographical comparator?

If so, perhaps, it should either be worked around (to somehow sort leaving last `is_ge()` value) or maybe it doesn't worth that, since this will work only for single threaded queries, while, I suppose, the primarily use-case is multithreaded execution, which messes up the original order of records in files due to parallel reads followed by repartitioning.

\+ Not sure if current behaviour of this function is incorrect and must be fixed, it looks more like nice-to-have feature (but that is just an opinion).

Reply:
Maybe we need benchmark on this.

With additional indices columns, we are able to optimize it with fetch logic, where we just take the last one after sorted. 

Without additional indices columns, we need to iterate them all the find the correct one, it includes the cost of comparison


> since this will work only for single threaded queries, while, I suppose, the primarily use-case is multithreaded execution, which messes up the original order of records in files due to parallel reads followed by repartitioning. 

I think even in this case, we still expect to return the correct number?

Reply:
Okay, it seems like unstable sort is used in arrow-rs probably there is no **fast** way to achieve it.

Alternative solution is that the users can add another column by themselves to get the **true** last value if they need for their scenario 🤔

Reply:
> I think even in this case, we still expect to return the correct number?

Don't think we can -- it depends on the order of partial aggregation stream completions, which it not really predictable, so function potentially may produce non-deterministic output in this case.

> Alternative solution is that the users can add another column by themselves to get the true last value if they need for their scenario

I'd say that this is how it works now, which seems to be good enough.

Also, maybe the initial version of this PR, was sufficient? It didn't affect performance, and it provided guarantees for `target_partitions=1` that the function will return the value which is placed last in source files.

Reply:
The initial PR doesn't work for single partition case too since the `get_last_idx` return the first (value,order) pair not the last one

Reply:
DuckDB also return the first one not the "true" last one

Reply:
> The initial PR doesn't work for single partition case too since the get_last_idx return the first (value,order) pair not the last one

Yes, my bad, lost the context for a while. Then ye, looks like providing unique column as an `order by` is the only way to make the output deterministic.